src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,884 @@
+		goto out;
+	}
+
+	rc = dfs_connect(pool, NULL, cont, hdl->flags, NULL, &hdl->dfs);
+	if (rc) {
+		D_ERROR("Could not connect to %s:%s: %s (rc=%d)", pool, cont, strerror(rc), rc);
+		goto out;
+	}
+
+	hdl->global.iov_buf     = NULL;
+	hdl->global.iov_buf_len = 0;
+
+	rc = dfs_local2global_all(hdl->dfs, &hdl->global);
+	if (rc) {
+		D_ERROR("Could not get global handler size for dfs: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	D_ALLOC(hdl->global.iov_buf, hdl->global.iov_buf_len);
+	if (hdl->global.iov_buf == NULL) {
+		rc = ENOMEM;
+		goto out;
+	}
+
+	rc = dfs_local2global_all(hdl->dfs, &hdl->global);
+	if (rc) {
+		D_ERROR("Could not create global handler for dfs: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	rc = daos_eq_create(&hdl->eq);
+	if (rc) {
+		D_WARN("Could not create event queue: %s (rc=%d)", d_errstr(rc), rc);
+		goto out;
+	}
+	hdl->eq_owner_pid = getpid();
+
+out:
+	if (rc) {
+		dfs_disconnect(hdl->dfs);
+
+		D_FREE(hdl->global.iov_buf);
+		D_FREE(hdl);
+		hdl = NULL;
+	}
+
+	PyList_SetItem(result, 0, PyLong_FromLong(rc));
+	PyList_SetItem(result, 1, PyLong_FromVoidPtr(hdl));
+
+	return result;
+}
+
+static PyObject *
+__shim_handle__torch_disconnect(PyObject *self, PyObject *args)
+{
+	struct dfs_handle *hdl = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "K", &hdl);
+
+	assert(hdl->dfs != NULL);
+
+	if (hdl->eq_owner_pid != getpid()) {
+		return PyLong_FromLong(EACCES);
+	}
+
+	int rc = dfs_disconnect(hdl->dfs);
+	if (rc) {
+		D_ERROR("Could not disconnect DFS: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	rc = daos_eq_destroy(hdl->eq, DAOS_EQ_DESTROY_FORCE);
+	if (rc) {
+		D_ERROR("Could not destroy event queue: %s (rc=%d)", d_errstr(rc), rc);
+		goto out;
+	}
+
+	/* DER_BUSY errors are due to module sharing via python multiprocessing.
+	 * The last process calling disconnect will free the resources */
+	rc = dfs_fini();
+	if (rc && rc != -DER_BUSY) {
+		D_ERROR("Could not finalise DFS: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	rc = daos_fini();
+	if (rc == -DER_BUSY) {
+		D_ERROR("Could not finalise DAOS: %s (rc=%d)", d_errstr(rc), rc);
+		rc = DER_SUCCESS;
+	}
+
+out:
+	D_FREE(hdl->global.iov_buf);
+	D_FREE(hdl);
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__torch_reinit(PyObject *self, PyObject *args)
+{
+	struct dfs_handle *hdl = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "K", &hdl);
+
+	if (hdl->eq_owner_pid == getpid()) {
+		return PyLong_FromLong(0);
+	}
+
+	int rc = dfs_init();
+	if (rc) {
+		return PyLong_FromLong(rc);
+	}
+
+	hdl->dfs          = NULL;
+	hdl->eq           = DAOS_HDL_INVAL;
+	hdl->eq_owner_pid = getpid();
+
+	rc = dfs_global2local_all(hdl->flags, hdl->global, &hdl->dfs);
+	if (rc) {
+		D_ERROR("Could not create local handler from global one: %s (rc=%d)", strerror(rc),
+			rc);
+		return PyLong_FromLong(rc);
+	}
+
+	rc = daos_eq_create(&hdl->eq);
+	if (rc) {
+		D_ERROR("Could not create event queue: %s (rc=%d)", d_errstr(rc), rc);
+	}
```

Comment:
I don't think you need that any longer with the new daos_reinit() API.

Reply:
Does not work for me without re-creating event queue in child process.